server/api.go

```diff
@@ -108,7 +109,17 @@ func (p *Plugin) attachUserContext(handler HTTPHandlerFuncWithUserContext) http.
 		context, cancel := p.createContext(w, r)
 		defer cancel()
 
+		mutex, err := cluster.NewMutex(p.API, context.UserID+"-refresh-token")
+		if err != nil {
+			p.writeAPIError(w, &APIErrorResponse{ID: "", Message: "Unable to get mutex: " + err.Error(), StatusCode: http.StatusInternalServerError})
+			return
+		}
+
+		mutex.Lock()
+
 		info, apiErr := p.getGitlabUserInfoByMattermostID(context.UserID)
+		defer mutex.Unlock()
```

Comment:
@mickmister It would be necessary to lock the entire handler if the call to `checkAndRefreshToken` is inside the handler, as it was up until my most recent change which moved it to the `checkToken` wrapper instead. So I could probably work it so `attachUserContext` and `checkToken` are combined somehow and the lock is released before the handler is called, except it sounds in your other comments like you are thinking of putting the token check logic back inside the handlers?

Yeah I'm not sure why this problem is so difficult, but I've been kind of going back and forth between these options myself: should the check be closer to the GitLab calls, or should it be closer to the origin of the request?